### PR TITLE
coretasks: better SASL handling

### PR DESCRIPTION
**Pull request by dgw**
### Description
Took a WIP fix for #1780 I found lying around the branches and extended it.

### Checklist
- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [ ] I have tested the functionality of the things this change touches

### Notes
Better late than never (just look how long this PR sat open…)

### CODE REVIEW

File: sopel/coretasks.py [hunk 1]
```diff
@@ -741,7 +742,17 @@ def receive_cap_ack_sasl(bot):
         mech = bot.config.core.server_auth_sasl_mech
     if not password:
         return
+
     mech = mech or 'PLAIN'
+    available_mechs = bot.server_capabilities.get('sasl', '')
+    available_mechs = available_mechs.split(',') if available_mechs else []
```
Comment (cottongin) [inline sopel/coretasks.py:748]: `bot.server_capabilities.get('sasl', []).split(',')` seems simpler here

File: sopel/coretasks.py [hunk 2]
```diff
@@ -741,7 +742,17 @@ def receive_cap_ack_sasl(bot):
...
+
+    if available_mechs and mech not in available_mechs:
```
Comment (cottongin) [inline sopel/coretasks.py:750]: Why the check for `available_mechs in available_mechs`? Am I missing something here?

Comment (dgw) [reply sopel/coretasks.py:750]: It's grouped like `if (available_mechs) and (mech not in available_mechs)`; see the comment below for why.

File: sopel/coretasks.py [hunk 3]
```diff
@@ -832,6 +850,46 @@ def sasl_success(bot, trigger):
...
+        1. Sopel connects to a network advertising SASL 3.1
+        2. SASL 3.1 doesn't advertise supported mechanisms up front, so Sopel
+           blindly goes ahead with whatever SASL config it's set to use
+        3. The server doesn't support the mechanism Sopel used, and is a good
+           IRC citizen, so it sends this optional numeric, 908 RPL_SASLMECHS
+
+        Note that misconfigured SASL 3.1 will just silently fail when connected
+        to an IRC server NOT implementing the optional 908 reply.
+        A network with SASL 3.2 should theoretically never get this far because
+
```
Comment (Exirel) [inline sopel/coretasks.py:882]:
```suggestion
```

**Review by Exirel — APPROVED** [review]
One nitpick for you to fix (unless there is a reason I'm not seeing), and also don't hesitate to change the PR's description. 👍

File: sopel/coretasks.py (hunk 3 above)
Comment (dgw) [reply sopel/coretasks.py:867]: I've no idea how that happened 👀; will fix.

Though now I've noticed this seven-year-old extra blank in `blocks()`: https://github.com/sopel-irc/sopel/blob/b16cfcd8e1197e4e85727f00f56c69f01c890aac/sopel/coretasks.py#L843-L848

**Comment by dgw** [comment]
@Exirel Yes, I saw your aside on IRC about the repeated SASL mech/password lookup code. What do you think now? 😁

File: sopel/coretasks.py [hunk 4]
```diff
@@ -832,6 +843,51 @@ def sasl_success(bot, trigger):
...
+def _get_sasl_pass_and_mech(bot):
+    password = None
+    mech = None
+    if bot.config.core.auth_method == 'sasl':
+        password = bot.config.core.auth_password
+        mech = bot.config.core.auth_target
+    elif bot.config.core.server_auth_method == 'sasl':
+        password = bot.config.core.server_auth_password
+        mech = bot.config.core.server_auth_sasl_mech
+    return password, mech
```
Comment (Exirel) [inline sopel/coretasks.py:888]:
```suggestion
def _get_sasl_pass_and_mech(settings):
    password = None
    mech = None
    if settings.core.auth_method == 'sasl':
        password = settings.core.auth_password
        mech = settings.core.auth_target
    elif settings.core.server_auth_method == 'sasl':
        password = settings.core.server_auth_password
        mech = settings.core.server_auth_sasl_mech
    return password, mech
```

File: sopel/coretasks.py [hunk 5]
```diff
@@ -832,6 +843,51 @@ def sasl_success(bot, trigger):
     bot.write(('CAP', 'END'))
 
 
+@module.event(events.RPL_SASLMECHS)
+def sasl_mechs(bot, trigger):
+    # Presumably we're only here if we said we actually *want* sasl, but still
+    # check anyway in case the server glitched.
+    password, mech = _get_sasl_pass_and_mech(bot)
```
Comment (Exirel) [inline sopel/coretasks.py:850]:
```suggestion
    password, mech = _get_sasl_pass_and_mech(bot.config)
```

File: sopel/coretasks.py [hunk 6]
```diff
@@ -762,18 +763,28 @@ def acct_warn(bot, cap):
...
-    password = None
-    mech = None
-    if bot.config.core.auth_method == 'sasl':
-        password = bot.config.core.auth_password
-        mech = bot.config.core.auth_target
-    elif bot.config.core.server_auth_method == 'sasl':
-        password = bot.config.core.server_auth_password
-        mech = bot.config.core.server_auth_sasl_mech
+    # check anyway in case the server glitched.
+    password, mech = _get_sasl_pass_and_mech(bot)
```
Comment (Exirel) [inline sopel/coretasks.py:767]: OK it's kind of a nitpick here: if all you need is the settings, then just provide the settings. I've added other suggestion to reflect that change, so you can commit them in one go.

```suggestion
    password, mech = _get_sasl_pass_and_mech(bot.config)
```

**Review by Exirel — APPROVED** [review]
One nitpick that doesn't prevent this PR to be accepted in my opinion. 👍

**Comment by dgw** [comment]
@Exirel Review re-requested because of the new commit making three SASL handlers "unblockable", which I noticed while browsing through `coretasks` again to look for something else.

About changing the helper to take `bot.config` as input instead of `bot`: Why not go all the way and pass `bot.config.core`? It certainly doesn't need any of the other sections. Where do we stop? 😁

**Comment by Exirel** [comment]
>  Where do we stop? 😁

We stop at the more accurate higher level of objects. The function wants the settings for SASL. Today it's in core, maybe tomorrow that'll be in core and another section, or only in another section, etc.

**Review by Exirel — APPROVED** [review]

**Comment by dgw** [comment]
> Today it's in core, maybe tomorrow that'll be in core and another section, or only in another section, etc.

Or maybe it'll be in `bot.db` next week. 😼

**Pull request merged**